feat: implement strimzi-ui build

matthew-chirgwin opened the pull request:
- Offers an implementation of the UI build detailed in #2 
- Adds extra documentation/scripts to allow easy running of the build
- Adds a UI (which will be replaced) which makes use of this build. All elements of the build bar fonts are demonstrated in this simple 'Hello world' UI.


Closes: #1

matthew-chirgwin committed:
feat: provide build and basic UI

- Adds implementation for the strimzi-ui build, inc npm package
and required dependancies
- Adds a very basic UI, which will be replaced, which uses this build
- Update documentation to cover

Contributes to: strimzi/strimzi-ui#2

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
refactor: build implementation details

- moves dependencies to correct section, depending if they or code they
generate are included in built output
- adds browser levels to babel presets
- adds webpack-bundle-analyzer for bundle size checking / docs to cover
- tweaks location and merging of configuration to better match docs
- adds extra comments/documentation

Contributes to: strimzi/strimzi-ui#2

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

samuel-hawker commented on .gitignore:
```diff
@@ -0,0 +1,22 @@
+# dependancies
```
Typo

matthew-chirgwin replied:
Good spot - will correct in next commit

matthew-chirgwin commented:
@nictownsend / @seanforyou23 would you mind reviewing this?

matthew-chirgwin committed:
refactor: spelling error, review comments

- fixes spelling errors

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
feat: add copyright banner to css

- adds copyright banner to built css

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend commented on README.md:
```diff
@@ -5,3 +5,32 @@
+This UI uses `npm` to provide dependency management. If you wish to develop the UI, you will need:
+
+- [npm version 6.4.1 or later](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
+- [node 10.15.0 or later](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm)
```
Can we fix to either latest or LTS? And then we can reflect that into the docker image as needed later on.

matthew-chirgwin replied:
These are the levels I personaly built the above with (which I know work). More than happy to bump these - worth an issue to do in your opinion to discuss more widley/implement the result?

nictownsend commented on client/Bootstrap/index.js:
```diff
@@ -0,0 +1,16 @@
+import image from 'Images/logo.png';
+
+window.addEventListener('load', () => {
+    const root = document.querySelector('[id="root"]');
```
I'm guessing this is just so that we have placeholder JS to prove webpack is bundling? And will later replace with React logic?

matthew-chirgwin replied:
As per the PR, yes, completley placeholder/Hello world, to use the build. All bar fonts have been utilised in this UI

nictownsend commented on docs/Build.md:
```diff
@@ -66,7 +66,7 @@ In addition to aliasing, the webpack configuration will be as follows:
 | devServer      | Object | [See section below](#webpack-dev-server) |
 
-This configuration will be defined in separate files per build mode - ie one for `development` and one for `production` - to keep the configuration as minimal and readable as possible. Common configuration will be inherited/extended into these files.
+This configuration will be defined in separate files per build mode - ie one for `development` and one for `production` - to keep the configuration as minimal and readable as possible. Common configuration will be inherited/extended into these files. [See this section for more details](#ui-build-implementation).
```
To keep the configuration as minimal and readable as possible, configuration will be defined in separate files per build mode - ie one for `development` and one for `production`. These configurations will extend a common configuration file.

nictownsend commented on docs/Build.md:
```diff
@@ -90,6 +90,8 @@ We also make use of the following plugins:
 | mini-css-extract-plugin      | Compresses and emits a css file(s) containing all styling for the UI |
 | compression-webpack-plugin      | Compresses built output further using `gzip` algorithm, and emits these compressed files. Depending on headers provided by the browser, these gziped versions will be served to the browser, rather than the uncompressed versions |
+| webpack-bundle-analyzer      | At build time, produce a report regarding the JS bundle size (useful for understanding bloat and duplication). At dev build time this is an html file (which is then hosted by `webpack-dev-server`), and a json file at production build time. Each report is written to the `generated/bundle-analyser` directory |
+| BannerPlugin      | Used to add a copyright header to built css code. Other types handled by other plugins (eg TerserPlugin) |
```
Any way we can get headers auto added to source?

matthew-chirgwin replied:
Yes - its in the linting implementation I have ready :)

samuel-hawker reviewed (approved):
Personally I am happy with this implementation.

Did any other non maintainers have any thoughts?

nictownsend commented on package.json:
```diff
@@ -0,0 +1,46 @@
+  },
+  "homepage": "https://github.com/strimzi/strimzi-ui#readme",
+  "dependencies": {
+    "@babel/core": "^7.10.5",
```
Are any of these dependencies? I'm guessing `express` or equivalent will be in the future - but for now, all of this is just to build assets into the `public` folder, none are needed for running the application.

(Unless we're using the semantics of "it produces production output", but then I'd say they all are dependencies)

matthew-chirgwin replied:
I havent written it down yet, but I am of the view that it is a production dependency if it produces output that are built. Those listed in dev currently are either developer dependencies, (ie not used in shipped code) or used in the build, but do not generate output.

nictownsend commented on utils/build/webpack.common.js:
```diff
@@ -0,0 +1,159 @@
+ */
+const configurationCombiner = (objValue, srcValue) => {
+    let combinedResult = undefined;
+    if (Array.isArray(objValue) && Array.isArray(srcValue)) {
```
https://lodash.com/docs/4.17.15#difference for combining?

matthew-chirgwin replied:
I shall look into this now :)

nictownsend commented on utils/build/webpack.dev.js:
```diff
@@ -0,0 +1,38 @@
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+const { returnWebpackConfigFor, plugins, CONSTANTS } = require('./webpack.common.js');
```
This feels a bit like overkill - could this be achieved by getting the common config and just using object spread to add in new properties?

matthew-chirgwin replied:
I don't think so (mainly due to how the config is defined in Webpack in arrays and how I want to define as much as possible which is common in that common file) - would be happy to ajust/update if you can come up with something

matthew-chirgwin committed:
refactor: simplify build based on review

- changes/simplifies the build mechanism to become more additive
- doc changes to cover

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend reviewed (approved):
I think this may need revisiting if the configs become more dynamic in future - i.e no consistent ordering to the modules.

But for now, good compromise to avoid duplication and complexity.

dlabaj reviewed (approved):
LGTM

matthew-chirgwin commented:
Thank you all - please if all are happy, please feel free to merge at the earliest convenience
